diff --git a/COFF/Writer.cpp b/COFF/Writer.cpp
index d17405ec26a..51f60bd3348 100644
--- a/COFF/Writer.cpp
+++ b/COFF/Writer.cpp
@@ -459,7 +459,7 @@ void Writer::createSections() {
     // sections are treated as if they have output characteristics DATA | R if
     // their characteristics are DATA | R | W. This implements the same special
     // case for all architectures.
-    if (Name == ".CRT")
+    if (Name == ".CRT" || Name == ".rsrc")
       OutChars = DATA | R;
 
     OutputSection *Sec = CreateSection(Name, OutChars);
@@ -975,8 +975,6 @@ void Writer::createSEHTable() {
   for (ObjFile *File : ObjFile::Instances) {
     // FIXME: We should error here instead of earlier unless /safeseh:no was
     // passed.
-    if (!File->hasSafeSEH())
-      return;
 
     markSymbolsForRVATable(File, File->getSXDataChunks(), Handlers);
   }
