diff --git a/lib/CodeGen/AsmPrinter/WinException.cpp b/lib/CodeGen/AsmPrinter/WinException.cpp
index eff73a58d8d..d905c79bab1 100644
--- a/lib/CodeGen/AsmPrinter/WinException.cpp
+++ b/lib/CodeGen/AsmPrinter/WinException.cpp
@@ -718,7 +718,7 @@ void WinException::emitCXXFrameHandler3Table(const MachineFunction *MF) {
   OS.EmitIntValue(0, 4);
 
   AddComment("EHFlags");
-  OS.EmitIntValue(1, 4);
+  OS.EmitIntValue(0, 4);
 
   // UnwindMapEntry {
   //   int32_t ToState;
